core/src/commonMain/kotlin/info/laht/threekt/controls/FirstPersonControls.kt

```diff
@@ -0,0 +1,240 @@
+		if (moveBackward) camera.translateZ(actualMoveSpeed)
+
+		if (moveLeft) camera.translateX(-actualMoveSpeed)
+		if (moveRight) camera.translateX(actualMoveSpeed)
+
+		if (moveUp) camera.translateY(-actualMoveSpeed)
+		if (moveDown) camera.translateY(actualMoveSpeed)
+
+		var actualLookSpeed = delta * lookSpeed
+
+		if (!activeLook) {
+
+			actualLookSpeed = 0.0f
+
+		}
+
+		var verticalLookRatio = 1.0
+
+		if (constrainVertical) {
+
+			verticalLookRatio = PI / (verticalMax - verticalMin)
+
+		}
+
+		lon -= (mouseX * actualLookSpeed).toInt()
+		if (lookVertical) lat -= (mouseY * actualLookSpeed * verticalLookRatio).toInt()
+
+		lat = max(-85, min(85, lat))
+
+		var phi = degToRad(90 - lat)
+		val theta = degToRad(lon)
+
+		if (constrainVertical) {
+
+			phi = mapLinear(phi, 0.0f, PI.toFloat(), verticalMin.toFloat(), verticalMax.toFloat())
+
+		}
+
+		val position = camera.position
+
+		targetPosition.setFromSphericalCoords(1.0f, phi, theta).add(position)
+
+		camera.lookAt(targetPosition)
+	}
+
+	fun dispose() {
+		eventSource.removeKeyListener(defaultKeyListener)
+		eventSource.removeMouseListener(defaultMouseListener)
+	}
+
+	fun setOrientation(controls: FirstPersonControls) {
+		val quaternion = controls.camera.quaternion
+
+		lookDirection.set(0, 0, -1).applyQuaternion(quaternion)
+		spherical.setFromVector3(lookDirection)
+
+		lat = (90 - radToDeg(spherical.phi)).toInt()
+		lon = radToDeg(spherical.theta).toInt()
+	}
+
+	private inner class MyKeyListener : KeyListener {
+		override fun onKeyPressed(event: KeyEvent) {
+			var needUpdate = true
+			when (event.keyCode) {
+				38, 87 -> moveForward = true
+				37, 65 -> moveLeft = true
+				40, 83 -> moveBackward = true
+				39, 68 -> moveRight = true
+				82 -> moveUp = true
+				70 -> moveDown = true
+				else -> needUpdate = false
+			}
+			if (needUpdate) {
+				update()
+			}
+		}
+
+		override fun onKeyReleased(event: KeyEvent) {
+			var needUpdate = false
+			when (event.keyCode) {
+				38, 87 -> moveForward = false
+				37, 65 -> moveLeft = false
+				40, 83 -> moveBackward = false
+				39, 68 -> moveRight = false
+				82 -> moveUp = false
+				70 -> moveDown = false
+				else -> needUpdate = false
+			}
+			if (needUpdate) {
+				update()
+			}
+		}
+	}
+
+	private inner class MyMouseListener : MouseAdapter() {
+		override fun onMouseDown(button: Int, event: MouseEvent) {
+			if (activeLook) {
+				when (button) {
+					0 -> moveForward = true //0 = left
+					1 -> moveBackward = true //1 == right
+				}
+			}
+			mouseDragOn = true
+		}
+
+		override fun onMouseUp(button: Int, event: MouseEvent) {
+			if (activeLook) {
+				when (button) {
+					0 -> moveForward = false
+					1 -> moveBackward = false
+				}
+			}
+			mouseDragOn = false
+		}
+
+		override fun onMouseMove(event: MouseEvent) {
+			mouseX = event.clientX - viewHalfX
+			mouseY = event.clientY - viewHalfY
+		}
+	}
+
+}
```

Comment:
Files should end with a newline. Goes for all files that GitHub marks with this symbols